ltm/agent.py

```diff
@@ -96,6 +125,121 @@ def sanitize_string(s: str) -> str:
+        "trigger_condition": str // A precise description of the condition that should be met for the task to execute.
+        "payload": str, // An exact description of what should be done when the trigger condition is met. Do not mention the trigger
+        // Any other details that might be relevant
+    }
+    "trigger_details": {
+        "trigger_condition": str, // Either MESSAGE_COUNT, MESSAGE_CONTENT, or TIME
+        "trigger_value": str, // For MESSAGE_COUNT, the number of messages to wait. For TIME, a timestamp in the form `YYYY-mm-dd_HH:MM`. For MESSAGE_CONTENT, the exact text that will trigger the task.
+    }
+}"""
+
+        context.append(make_user_message(task_should_add.format(future_tasks=json.dumps(self.task_memory, indent=2), user_message=user_message)))
+        response_text = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+        log_llm_call(self.run_name, self.save_name, self.debug_level, label=f"should_update-{self.llm_call_idx}")
+        colour_print("YELLOW", f"Should update: {response_text}")
+
+        if "Should Update: No." in response_text:
+            return
+        elif "Should Update: Yes." not in response_text:
+            return
+
+        # Produce a new one
+        context.append(make_assistant_message(response_text))
+        context.append(make_user_message(new_task_def))
+
+        try:
+            response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+            log_llm_call(self.run_name, self.save_name, self.debug_level, label=f"do_update-{self.llm_call_idx}")
+
+            new_task = sanitize_and_parse_json(response)
+            self.process_new_task(new_task)
+
+            print(f"Added Task:\n{json.dumps(new_task, indent=2)}\n")
+
+        except Exception as e:
+            print("Encountered exception when creating task: " + repr(e))
+
+    def process_new_task(self, new_task: dict):
+        new_task["trigger_details"]["triggered"] = False
+        if new_task["trigger_details"]["trigger_condition"] == "MESSAGE_COUNT":
+            new_task["trigger_details"]["trigger_value"] = int(new_task["trigger_details"]["trigger_value"])
+
+        if new_task["trigger_details"]["trigger_condition"] == "MESSAGE_CONTENT":
+            new_task["trigger_details"]["trigger_value"] = new_task["trigger_details"]["trigger_value"].lower()
+
+        if new_task["trigger_details"]["trigger_condition"] == "TIME":
+            new_task["trigger_details"]["trigger_value"] = time.mktime(datetime.datetime.strptime(new_task["trigger_details"]["trigger_value"], "%Y-%m-%d_%H:%M").timetuple())
+
+        self.task_memory.append(new_task)
+
+    def remove_completed_tasks(self):
+
+        for task in self.task_memory:
+            if task["trigger_details"]["triggered"]:
+                self.task_memory.remove(task)
```

Comment:
`remove` looks for a perfect match, and `task` is a dictionary with many details. A safer way would be:
```python
self.task_memory = [
    t for t in self.task_memory
    if not t["trigger_details"]["triggered"]
]
```